apps/shared/avifpng.c

```diff
@@ -162,75 +138,63 @@ static avifBool avifExtractExifAndXMP(png_structp png, png_infop info, avifBool
             *ignoreExif = AVIF_TRUE; // Ignore any other Exif chunk.
         } else if (!*ignoreXMP && !strcmp(text->key, "Raw profile type xmp")) {
-            if (!avifCopyRawProfile(text->text, avifGetPngTextLength(text), &metadata, &tagExif00WasRemoved)) {
+            if (!avifCopyRawProfile(text->text, textLength, &avif->xmp)) {
                 return AVIF_FALSE;
             }
-            if (tagExif00WasRemoved) {
-                avifRWDataFree(&metadata);
-                fprintf(stderr, "XMP extraction failed: XMP payload begins with an Exif tag\n");
-                return AVIF_FALSE;
-            }
-            avifRWDataFree(&avif->xmp);
-            avif->xmp = metadata;
             *ignoreXMP = AVIF_TRUE; // Ignore any other XMP chunk.
         } else if (!strcmp(text->key, "Raw profile type APP1")) {
-            if (!avifCopyRawProfile(text->text, avifGetPngTextLength(text), &metadata, &tagExif00WasRemoved)) {
+            // This can be either Exif, XMP or something else.
+            avifRWData metadata = { NULL, 0 };
+            if (!avifCopyRawProfile(text->text, textLength, &metadata)) {
                 return AVIF_FALSE;
             }
-            // Differentiate Exif and XMP by checking whether the chunk begins with a valid TIFF header ("IIx0x42" or "MMx42x0").
-            const uint32_t tiffHeaderSize = 4;
+            const uint32_t exifTiffHeaderSize = 10; // "MM\0\42" or "II\42\0", maybe prefixed by "Exif\0\0"
             uint32_t exifTiffHeaderOffset;
-            if (metadata.size > tiffHeaderSize &&
-                (avifExtractExifTiffHeaderOffset(metadata.data, tiffHeaderSize, &exifTiffHeaderOffset) == AVIF_RESULT_OK)) {
+            const avifROData xmpJpegHeader = { (const uint8_t *)"http://ns.adobe.com/xap/1.0/\0", 29 };
+            if ((metadata.size > exifTiffHeaderSize) &&
+                (avifExtractExifTiffHeaderOffset(metadata.data, exifTiffHeaderSize, &exifTiffHeaderOffset) == AVIF_RESULT_OK)) {
+                // metadata has an exif header.
                 if (*ignoreExif) {
                     avifRWDataFree(&metadata);
                 } else {
                     avifRWDataFree(&avif->exif);
                     avif->exif = metadata;
+                    avifRemoveHeadingExif00TagIfAny(&avif->exif);
                     *ignoreExif = AVIF_TRUE; // Ignore any other Exif chunk.
                 }
-            } else if (tagExif00WasRemoved) {
-                if (*ignoreExif) {
-                    // Ignore invalid Exif payloads too.
-                    avifRWDataFree(&metadata);
-                } else {
-                    avifRWDataFree(&metadata);
-                    fprintf(stderr, "Exif extraction failed: missing TIFF header\n");
-                    return AVIF_FALSE;
-                }
-            } else {
-                // For simplicity, consider everything else as XMP.
+            } else if ((metadata.data[0] == '<') ||
```

Comment:
Don't think there is point in checking for "<" - in the APP1 case the payload must start w/ the header above.